### PR TITLE
log errors on metric server and gracefully terminate metrics endpoint

### PR DESCRIPTION
**Pull request by rphillips**
#### What type of PR is this?
/bug

#### What this PR does / why we need it:
logrus.Fatalf will cause crio to exit with a non-zero return code. This patch changes the fatal error to an error preventing crio from exiting with a (1) status code. The metrics server is gracefully shutdown on termination.

#### Does this PR introduce a user-facing change?

```release-note
Change metrics endpoint to gracefully shutdown, and move metric server fatal errors to just errors. 
```

### CODE REVIEW

**Review by cgwalters — APPROVED** [review]

**Comment by haircommander** [comment]
should we retry for a period instead? I worry this will break more things if cri-o never tries to reestablish the metrics connection

File: server/metrics/metrics.go [hunk 1]
```diff
@@ -596,51 +588,54 @@ func (m *Metrics) createEndpoint() (*http.ServeMux, error) {
...
+			continue
+		}
+
+		srv := &http.Server{
+			Handler: me,
+		}
+
+		go func() {
+			<-stop
+			stopped = true
```
Comment (TomSweeneyRedHat) [inline server/metrics/metrics.go:607]: should this be after the if statement?  If the shutdown fails, you're not stopped are you?

File: server/metrics/metrics.go [hunk 2]
```diff
@@ -596,51 +588,54 @@ func (m *Metrics) createEndpoint() (*http.ServeMux, error) {
...
+			if err := srv.Shutdown(context.Background()); err != nil {
+				logrus.Warnf("error in metrics endpoint shutdown: %v", err)
```
Comment (mrunalp) [inline server/metrics/metrics.go:609]: nit: `error` -> `Error`. One of the CI tests is failing on this.

**Comment by mrunalp** [comment]
Looks fine besides one nit.

File: server/metrics/metrics.go [hunk 3]
```diff
@@ -596,51 +588,54 @@ func (m *Metrics) createEndpoint() (*http.ServeMux, error) {
...
-	if err != nil {
-		return fmt.Errorf("creating listener: %w", err)
-	}
+) {
+	stopped := false
+	for !stopped {
+		l, err := net.Listen(network, address)
+		if err != nil {
+			logrus.Warnf("Failed to create listener %v", err)
+			time.Sleep(15 * time.Second)
```
Comment (haircommander) [inline server/metrics/metrics.go:597]: same here with the sleep

File: server/metrics/metrics.go [hunk 4]
```diff
@@ -596,51 +588,54 @@ func (m *Metrics) createEndpoint() (*http.ServeMux, error) {
...
 func (m *Metrics) startEndpoint(
 	stop chan struct{}, network, address string, me http.Handler,
-) error {
-	l, err := net.Listen(network, address)
-	if err != nil {
-		return fmt.Errorf("creating listener: %w", err)
-	}
+) {
+	stopped := false
+	for !stopped {
```
[the remark this answers is not in the document]
Comment (rphillips) [reply server/metrics/metrics.go:593]: it doesn't the variable is written in the same goroutine.

**Comment by rphillips** [comment]
Peter and I chatted on google meet. Changed the PR to be a bit more specific with less logic changes.

**Comment by haircommander** [comment]
/approve
LGTM

**Pull request merged**